{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "import os, pandas, re, lxml.etree, datetime, numpy, warnings, time\n",
    "\n",
    "pandas.options.display.max_rows=100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Date\n",
    "d = datetime.datetime.today()\n",
    "d = str(d.year)+\"_\"+str(d.month)+\"_\"+str(d.day)\n",
    "#d = \"2018_4_24\"\n",
    "\n",
    "wd_base = \"/Users/kf/Dropbox/kfdata/02_Data/my_db/Ensembl/release-91/sra/\"\n",
    "wd = wd_base+d\n",
    "\n",
    "# Set directory\n",
    "if not os.path.exists(wd):\n",
    "    os.mkdir(wd)\n",
    "os.chdir(wd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_search_term(species_name=\"\", bioprojects=[], biosamples=[]):\n",
    "    species_term = '\"'+species_name+'\"'+\"[organism]\"\n",
    "    tissues = [\"brain\", \"liver\", \"kidney\", \"testis\", \"testes\", \"heart\", \"cerebellum\", \"lung\", \"muscle\", \"spleen\", \"ovary\", \"colon\", \"thymus\", \"bone marrow\", \"lymph node\", \"skin\", \"blood\", \"adrenal\", \"placenta\", \"salivary gland\", \"uterus\", \"retina\", \"pancreas\", \"embryo\"]\n",
    "    tissue_term = \"(\" + \" OR \".join(tissues) + \")\"\n",
    "    other_conditions = ['\"platform illumina\"[Properties]', '\"type rnaseq\"[Filter]', '\"sra biosample\"[Filter]']\n",
    "    other_term = \" AND \".join(other_conditions)\n",
    "    excluded_conditions = ['\"strategy mirna seq\"[Properties]', '\"strategy rip seq\"[Properties]']\n",
    "    excluded_term = \" NOT \".join(excluded_conditions)\n",
    "    if len(bioprojects) and len(biosamples):\n",
    "        warnings.warn(\"Both bioprojects and biosamples cannot be specified.\")\n",
    "    elif len(bioprojects):\n",
    "        bioproject_term = \"(\" + \" OR \".join(bioprojects) + \")\"\n",
    "        search_term = species_term + \" AND \" + bioproject_term + \" AND \" + other_term + \" NOT \" + excluded_term\n",
    "    elif len(biosamples):\n",
    "        biosample_term = \"(\" + \" OR \".join(biosamples) + \")\"\n",
    "        search_term = biosample_term\n",
    "    else:        \n",
    "        search_term = species_term + \" AND \" + tissue_term + \" AND \" + other_term + \" NOT \" + excluded_term\n",
    "    return search_term\n",
    "\n",
    "def fetch_sra_xml(species_name, search_term, save_xml=True, read_from_existing_file=False):\n",
    "    file_xml = \"SRA_\"+species_name.replace(\" \", \"_\")+\".xml\"\n",
    "    flag = True\n",
    "    if (read_from_existing_file)&(os.path.exists(file_xml)):\n",
    "        with open(file_xml) as f:\n",
    "            if '<Error>' in f.read():\n",
    "                print(species_name, ': <Error> found in the saved file. Deleting...')\n",
    "                os.remove(file_xml)\n",
    "            else:\n",
    "                print(species_name, ': reading xml from file')\n",
    "                root = lxml.etree.parse(file_xml, parser=lxml.etree.XMLParser())\n",
    "                flag = False\n",
    "    if flag:\n",
    "        sra_handle = Entrez.esearch(db=\"sra\", term=search_term, retmax=10000000)\n",
    "        sra_record = Entrez.read(sra_handle)\n",
    "        record_ids = sra_record[\"IdList\"]\n",
    "        num_record = len(record_ids)\n",
    "        retmax = 1000\n",
    "        start_time = time.time()\n",
    "        query_search_time = datetime.datetime.now().strftime('%Y/%m/%d %H:%M:%S')\n",
    "        root = None\n",
    "        for i in numpy.arange(numpy.ceil(num_record//retmax)+1):\n",
    "            start = int(i*retmax)\n",
    "            end = int(((i+1)*retmax)-1) if num_record >= int(((i+1)*retmax)-1) else num_record\n",
    "            print('processing ', start, '-', end, flush=True)\n",
    "            handle = Entrez.efetch(db=\"sra\", id=record_ids[start:end], rettype=\"full\", retmode=\"xml\", retmax=retmax)\n",
    "            chunk = lxml.etree.parse(handle).getroot()\n",
    "            if root is None:\n",
    "                root = chunk\n",
    "            else:\n",
    "                root.append(chunk)\n",
    "        elapsed_time = int(time.time() - start_time)\n",
    "        xml_string = lxml.etree.tostring(root, pretty_print=True)\n",
    "        for line in str(xml_string).split('\\n'):\n",
    "            if '<Error>' in line:\n",
    "                print(line)\n",
    "                if os.path.exists(file_xml):\n",
    "                    os.remove(file_xml)\n",
    "                raise Exception(species_name, ': <Error> found in the xml.')\n",
    "        if save_xml:\n",
    "            with open(file_xml, 'wb') as f:\n",
    "                f.write(xml_string)\n",
    "    return root\n",
    "\n",
    "def sra_xml2table(species_name, xml_root):\n",
    "    if isinstance(xml_root, lxml.etree._Element):\n",
    "        xml_root = lxml.etree.ElementTree(xml_root)\n",
    "    root = xml_root\n",
    "    if isinstance(root, lxml.etree._ElementTree):\n",
    "        sra_table = pandas.DataFrame()\n",
    "        for entry in root.iter(tag=\"EXPERIMENT_PACKAGE\"):\n",
    "            items = []\n",
    "            bioproject = entry.findall('.//EXTERNAL_ID[@namespace=\"BioProject\"]')\n",
    "            if not len(bioproject):\n",
    "                labels = entry.findall('.//LABEL')\n",
    "                for label in labels:\n",
    "                    text = label.text\n",
    "                    if text.startswith(\"PRJ\"):\n",
    "                        bioproject = [label]\n",
    "                        break\n",
    "            is_single = len(entry.findall('.//LIBRARY_LAYOUT/SINGLE'))\n",
    "            is_paired = len(entry.findall('.//LIBRARY_LAYOUT/PAIRED'))\n",
    "            if is_single:\n",
    "                library_layout = [\"single\"]\n",
    "            elif is_paired:\n",
    "                library_layout = [\"paired\"]\n",
    "            else:\n",
    "                library_layout = [\"\"]\n",
    "            values = entry.findall('.//VALUE')\n",
    "            is_protected = [\"No\"]\n",
    "            if len(values):\n",
    "                for value in values:\n",
    "                    text = value.text\n",
    "                    if not text is None:\n",
    "                        if text.endswith(\"PROTECTED\"):\n",
    "                            is_protected = [\"Yes\"]\n",
    "                            break\n",
    "            items.append([\"bioproject\", bioproject])\n",
    "            items.append([\"scientific_name\", entry.xpath('./SAMPLE/SAMPLE_NAME/SCIENTIFIC_NAME')])\n",
    "            items.append([\"biosample\", entry.findall('.//EXTERNAL_ID[@namespace=\"BioSample\"]')])\n",
    "            items.append([\"experiment\", entry.xpath('./EXPERIMENT/IDENTIFIERS/PRIMARY_ID')])\n",
    "            items.append([\"run\", entry.xpath('./RUN_SET/RUN/IDENTIFIERS/PRIMARY_ID')])\n",
    "            items.append([\"sra_primary\", entry.xpath('./SUBMISSION/IDENTIFIERS/PRIMARY_ID')])\n",
    "            items.append([\"sra_sample\", entry.xpath('./SAMPLE/IDENTIFIERS/PRIMARY_ID')])\n",
    "            items.append([\"sra_study\", entry.xpath('./EXPERIMENT/STUDY_REF/IDENTIFIERS/PRIMARY_ID')])\n",
    "            items.append([\"published_date\", entry.xpath('./RUN_SET/RUN/@published')])\n",
    "            items.append([\"exp_title\", entry.xpath('./EXPERIMENT/TITLE')])\n",
    "            items.append([\"design\", entry.xpath('./EXPERIMENT/DESIGN/DESIGN_DESCRIPTION')])\n",
    "            items.append([\"lib_name\", entry.xpath('./EXPERIMENT/DESIGN/LIBRARY_DESCRIPTOR/LIBRARY_NAME')])\n",
    "            items.append([\"lib_strategy\", entry.xpath('./EXPERIMENT/DESIGN/LIBRARY_DESCRIPTOR/LIBRARY_STRATEGY')])\n",
    "            items.append([\"lib_source\", entry.xpath('./EXPERIMENT/DESIGN/LIBRARY_DESCRIPTOR/LIBRARY_SOURCE')])\n",
    "            items.append([\"lib_selection\", entry.xpath('./EXPERIMENT/DESIGN/LIBRARY_DESCRIPTOR/LIBRARY_SELECTION')])\n",
    "            items.append([\"lib_layout\", library_layout])\n",
    "            items.append([\"nominal_length\", entry.xpath('./EXPERIMENT/DESIGN/LIBRARY_DESCRIPTOR/LIBRARY_LAYOUT/PAIRED/@NOMINAL_LENGTH')])\n",
    "            items.append([\"nominal_sdev\", entry.xpath('./EXPERIMENT/DESIGN/LIBRARY_DESCRIPTOR/LIBRARY_LAYOUT/PAIRED/@NOMINAL_SDEV')])\n",
    "            items.append([\"spot_length\", entry.xpath('./EXPERIMENT/DESIGN/SPOT_DESCRIPTOR/SPOT_DECODE_SPEC/SPOT_LENGTH')])\n",
    "            items.append([\"read_index\", entry.xpath('./EXPERIMENT/DESIGN/SPOT_DESCRIPTOR/SPOT_DECODE_SPEC/READ_SPEC/READ_INDEX')])\n",
    "            items.append([\"read_class\", entry.xpath('./EXPERIMENT/DESIGN/SPOT_DESCRIPTOR/SPOT_DECODE_SPEC/READ_SPEC/READ_CLASS')])\n",
    "            items.append([\"read_type\", entry.xpath('./EXPERIMENT/DESIGN/SPOT_DESCRIPTOR/SPOT_DECODE_SPEC/READ_SPEC/READ_TYPE')])\n",
    "            items.append([\"base_coord\", entry.xpath('./EXPERIMENT/DESIGN/SPOT_DESCRIPTOR/SPOT_DECODE_SPEC/READ_SPEC/BASE_COORD')])\n",
    "            items.append([\"instrument\", entry.xpath('./EXPERIMENT/PLATFORM/ILLUMINA/INSTRUMENT_MODEL')])\n",
    "            items.append([\"lab\", entry.xpath('./SUBMISSION/@lab_name')])\n",
    "            items.append([\"center\", entry.xpath('./SUBMISSION/@center_name')])\n",
    "            items.append([\"submitter_id\", entry.xpath('./SUBMISSION/IDENTIFIERS/SUBMITTER_ID')])\n",
    "            items.append([\"study_title\", entry.xpath('./STUDY/DESCRIPTOR/STUDY_TITLE')])\n",
    "            items.append([\"pubmed_id\", entry.xpath('./STUDY/STUDY_LINKS/STUDY_LINK/XREF_LINK/ID')])\n",
    "            items.append([\"sample_title\", entry.xpath('./SAMPLE/TITLE')])\n",
    "            items.append([\"taxid\", entry.xpath('./SAMPLE/SAMPLE_NAME/TAXON_ID')])\n",
    "            items.append([\"sample_description\", entry.xpath('./SAMPLE/DESCRIPTION')])\n",
    "            items.append([\"total_spots\", entry.xpath('./RUN_SET/RUN/@total_spots')])\n",
    "            items.append([\"total_bases\", entry.xpath('./RUN_SET/RUN/@total_bases')])\n",
    "            items.append([\"size\", entry.xpath('./RUN_SET/RUN/@size')])\n",
    "            items.append([\"is_protected\", is_protected])\n",
    "            row = []\n",
    "            for item in items:\n",
    "                try:\n",
    "                    if isinstance(item[1][0], (lxml.etree._ElementUnicodeResult, int, str)):\n",
    "                        row.append(str(item[1][0]))\n",
    "                    else:\n",
    "                        row.append(item[1][0].text)\n",
    "                except:\n",
    "                    row.append(\"\")\n",
    "            try:\n",
    "                column_names\n",
    "            except:\n",
    "                column_names = []\n",
    "                for item in items:\n",
    "                    column_names.append(item[0])\n",
    "            row_df = pandas.DataFrame(row).T\n",
    "            row_df.columns = column_names\n",
    "            sas = entry.xpath('./SAMPLE/SAMPLE_ATTRIBUTES/SAMPLE_ATTRIBUTE')\n",
    "            for sa in sas:\n",
    "                tag = sa.xpath('./TAG')\n",
    "                if not tag[0].text == None:\n",
    "                    tag = tag[0].text.lower()\n",
    "                    tag = re.sub(\" \\(.*\", \"\", tag)\n",
    "                    tag = re.sub(\" \", \"_\", tag)\n",
    "                    if not tag in row_df.columns:\n",
    "                        value = sa.xpath('./VALUE')\n",
    "                        if len(value):\n",
    "                            value = value[0].text\n",
    "                            if tag in column_names:\n",
    "                                tag = tag+\"_2\"\n",
    "                            sa_df = pandas.DataFrame([value])\n",
    "                            sa_df.columns = [tag]\n",
    "                            row_df = pandas.concat([row_df,sa_df], axis=1)\n",
    "            sra_table = pandas.concat([sra_table, row_df], ignore_index=True)\n",
    "        sra_table == sra_table.fillna(\"\")\n",
    "        if \"scientific_name\" in sra_table.columns and len(sra_table.loc[sra_table[\"scientific_name\"] == \"\", \"scientific_name\"]):        \n",
    "            sra_table.loc[sra_table[\"scientific_name\"] == \"\", \"scientific_name\"] = species_name\n",
    "        return sra_table\n",
    "    else:\n",
    "        raise Exception(\"Unknown input type.\", type(root))\n",
    "\n",
    "def exclude_by_id(sra_table, bioprojects=[], biosamples=[]):\n",
    "    print('Running: exclude_by_id')\n",
    "    if not 'exclusion' in sra_table.columns:\n",
    "        sra_table['exclusion'] = 'no'\n",
    "    sra_table.loc[sra_table['bioproject'].isin(bioprojects),'exclusion'] = 'bioproject'\n",
    "    sra_table.loc[sra_table['biosample'].isin(biosamples),'exclusion'] = 'biosample'\n",
    "    return sra_table\n",
    "\n",
    "def standardize_orthographical_variants(sra_table):\n",
    "    print('Running: standardize_orthographical_variants')\n",
    "    sra_table = sra_table.replace('not applicable', '')\n",
    "    sra_table = sra_table.replace('Not applicable', '')\n",
    "    sra_table = sra_table.replace('Not Applicable', '')\n",
    "    sra_table = sra_table.replace('not collected', '')  \n",
    "    sra_table = sra_table.replace('Missing', '') \n",
    "    sra_table = sra_table.replace('missing', '')\n",
    "    return sra_table\n",
    "\n",
    "def column_aggregation(sra_table, column_aggregation_list):\n",
    "    print('Running: column_aggregation')\n",
    "    for item in column_aggregation_list.items():\n",
    "        aggregate_to = item[0]\n",
    "        for aggregate_from in item[1]:\n",
    "            if (aggregate_from in sra_table.columns)&(aggregate_from!=''):\n",
    "                if not aggregate_to in sra_table.columns:\n",
    "                    sra_table[aggregate_to] = ''\n",
    "                is_from_empty = (sra_table[aggregate_from].isnull())|(sra_table[aggregate_from].astype(str)=='')\n",
    "                is_to_empty = (sra_table[aggregate_to].isnull())|(sra_table[aggregate_to].astype(str)=='')\n",
    "                new_annotations = sra_table.loc[(~is_from_empty)&(is_to_empty), aggregate_from].astype(str)+'['+aggregate_from+']'\n",
    "                sra_table.loc[(~is_from_empty)&(is_to_empty), aggregate_to] = new_annotations\n",
    "                new_annotations = sra_table.loc[(~is_from_empty)&(~is_to_empty), aggregate_to].astype(str)+\"; \"+sra_table.loc[(~is_from_empty)&(~is_to_empty), aggregate_from].astype(str)+'['+aggregate_from+']'\n",
    "                sra_table.loc[(~is_from_empty)&(~is_to_empty), aggregate_to] = new_annotations\n",
    "                sra_table = sra_table.drop(aggregate_from, 1)\n",
    "    sra_table = reorder_columns(sra_table, omit_misc=False)\n",
    "    return sra_table\n",
    "\n",
    "def standardize_tissue(sra_table, tissue_aggregation_list):\n",
    "    print('Running: standardize_tissue')\n",
    "    sra_table['tissue_original'] = sra_table['tissue']\n",
    "    sra_table['tissue'] = sra_table['tissue'].map(lambda x: str(x).lower())\n",
    "    sra_table['tissue'] = sra_table['tissue'].map(lambda x: re.sub(r'[0-9]+', '', str(x)))\n",
    "    sra_table['tissue'] = sra_table['tissue'].map(lambda x: re.sub('-', ' ', str(x)))\n",
    "    sra_table['tissue'] = sra_table['tissue'].map(lambda x: re.sub('_', ' ', str(x)))\n",
    "    sra_table['tissue'] = sra_table['tissue'].map(lambda x: re.sub(' \\(.*', '', str(x)))\n",
    "    sra_table['tissue'] = sra_table['tissue'].map(lambda x: re.sub('pouch young ', '', str(x)))   \n",
    "    sra_table[\"tissue\"] = sra_table[\"tissue\"].map(lambda x: re.sub(\"large white \", \"\", str(x)))\n",
    "    sra_table[\"tissue\"] = sra_table[\"tissue\"].map(lambda x: re.sub(\"pietrain \", \"\", str(x)))\n",
    "    sra_table[\"tissue\"] = sra_table[\"tissue\"].map(lambda x: re.sub(\"duroc \", \"\", str(x)))\n",
    "    sra_table[\"tissue\"] = sra_table[\"tissue\"].map(lambda x: re.sub(\"cherry valley duck \", \"\", str(x)))\n",
    "    sra_table['tissue'] = sra_table['tissue'].replace(';.*','',regex=True)\n",
    "    sra_table['tissue'] = sra_table['tissue'].replace(':.*','',regex=True)\n",
    "    sra_table['tissue'] = sra_table['tissue'].replace('\\[.*','',regex=True)\n",
    "    sra_table['tissue'] = sra_table['tissue'].replace('^ ','',regex=True)\n",
    "    sra_table['tissue'] = sra_table['tissue'].replace('^ ','',regex=True)\n",
    "    sra_table['tissue'] = sra_table['tissue'].replace(' $','',regex=True)\n",
    "    sra_table['tissue'] = sra_table['tissue'].replace(' $','',regex=True)\n",
    "    sra_table.loc[sra_table.tissue=='nan','tissue'] = ''\n",
    "    for item in tissue_aggregation_list.items():\n",
    "        to_value = item[0]\n",
    "        from_values = [ i for i in item[1] if i!='' ]\n",
    "        for from_value in from_values:\n",
    "            if from_value in list(sra_table['tissue']):\n",
    "                sra_table['tissue'] = sra_table['tissue'].replace(from_value, to_value)\n",
    "    return sra_table\n",
    "\n",
    "def standardize_sex(sra_table):\n",
    "    print(\"Running: standardize_sex\")\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: re.sub(\"female\", \"F\", str(x)))\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: re.sub(\"male\", \"M\", str(x)))\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: re.sub(\"FEM\", \"F\", str(x)))\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: re.sub(\"MALE\", \"M\", str(x)))\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: re.sub(\"Male\", \"M\", str(x)))\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: re.sub(\"Female\", \"F\", str(x)))\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: re.sub(\"fem\", \"F\", str(x)))\n",
    "    sra_table[\"sex\"] = sra_table[\"sex\"].map(lambda x: str(x).upper())\n",
    "    return sra_table\n",
    "\n",
    "def standardize_scientific_name(sra_table):\n",
    "    print(\"Running: standardize_scientific_name\")\n",
    "    sra_table[\"sci_name_original\"] = sra_table[\"scientific_name\"]\n",
    "    sra_table[\"scientific_name\"] = sra_table[\"scientific_name\"].replace(\"Canis familiaris\", \"Canis lupus\")\n",
    "    sra_table[\"scientific_name\"] = sra_table[\"scientific_name\"].replace(\"Xenopus (Silurana) tropicalis\", \"Xenopus tropicalis\") \n",
    "    sra_table[\"scientific_name\"] = [re.sub(r'(.+)(\\s)(.+)(\\s)(.+)', r\"\\1\\2\\3\", sp) for sp in sra_table[\"scientific_name\"]]\n",
    "    return sra_table\n",
    "\n",
    "def manual_attribute_filling(sra_table):\n",
    "    print(\"Running: manual_attribute_filling\")\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA215077\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA215077\", \"lib_name\"].map(lambda x: re.sub(\"cja_\", \"\", str(x)))\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA157897\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA157897\", \"lib_name\"].map(lambda x: re.sub(\"Beagle \", \"\", str(x)))\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA177703\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA177703\", \"exp_title\"].map(lambda x: re.sub(\"Canine \", \"\", str(x)))\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJDB3843\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJDB3843\", \"age\"]\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB11491\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB11491\", \"cell\"]\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA70959\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA70959\", \"exp_title\"]    \n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB7620\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB7620\", \"cell\"]\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA403803\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA403803\", \"source_name\"]\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA391214\", \"treatment\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA391214\", \"sample_title\"]\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB19268\", \"tissue\"] = sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB19268\", \"sample_title\"].map(lambda x: re.sub(\" RNA-seq\", \"\", str(x)))\n",
    "\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA173328\", \"tissue\"] = [\"uterus\", \"uterus\", \"ovary\"]\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJDB90\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA417542\", \"tissue\"] = \"vascular smooth muscle cells\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA222780\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJDB2921\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA209394\", \"tissue\"] = \"spleen\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJDB1766\", \"tissue\"] = \"cerebellum\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB12300\", \"tissue\"] = \"blood\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA189967\", \"tissue\"] = \"blood\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA168072\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB24166\", \"tissue\"] = \"macrophage\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB21709\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA188394\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB7406\", \"tissue\"] = \"spleen\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB22373\", \"tissue\"] = \"macrophage\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJNA341964\", \"tissue\"] = \"spleen\"\n",
    "    sra_table.loc[sra_table[\"bioproject\"]==\"PRJEB12613\", \"tissue\"] = \"skeletal muscle\"\n",
    "    \n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00706768\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00709563\", \"tissue\"] = \"thymus\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00727963\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00739297\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00739310\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00739311\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00739312\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00739313\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01766804\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01766813\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768054\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768055\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768056\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768057\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768058\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768059\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768060\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768061\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768062\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768063\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768064\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768065\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01768066\", \"tissue\"] = \"prefrontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00715283\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00715284\", \"tissue\"] = \"retina\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00778253\", \"tissue\"] = \"frontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00778254\", \"tissue\"] = \"frontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00778255\", \"tissue\"] = \"frontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00778256\", \"tissue\"] = \"frontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00778257\", \"tissue\"] = \"frontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00778258\", \"tissue\"] = \"frontal cortex\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00216499\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00216500\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00216504\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00216505\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00216506\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00216507\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00216508\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02401346\", \"tissue\"] = \"bone marrow\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02427293\", \"tissue\"] = \"temporal lobe\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02427294\", \"tissue\"] = \"pituitary\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02427295\", \"tissue\"] = \"lung\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02427296\", \"tissue\"] = \"skeletal muscle\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02427297\", \"tissue\"] = \"thymus\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00749950\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761040\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761041\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761045\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761046\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761050\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761051\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761052\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761053\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761054\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761055\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761056\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00761057\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01120728\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01120729\", \"tissue\"] = \"placenta\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01121009\", \"tissue\"] = \"blood\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01766975\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00991521\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00991522\", \"tissue\"] = \"spleen\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00991523\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00991525\", \"tissue\"] = \"ascending colon\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00991526\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00991527\", \"tissue\"] = \"lung\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00991528\", \"tissue\"] = \"cerebellum\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00765498\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00765499\", \"tissue\"] = \"ovary\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00765500\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00767849\", \"tissue\"] = \"skeletal muscle\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00767850\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00767851\", \"tissue\"] = \"lung\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00771449\", \"tissue\"] = \"embryo\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00791550\", \"tissue\"] = \"adrenal gland\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00014283\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00013560\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00013366\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00016407\", \"tissue\"] = \"lung/trachea\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00014285\", \"tissue\"] = \"pancreas\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00013365\", \"tissue\"] = \"spleen\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00013362\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00014284\", \"tissue\"] = \"lung/trachea\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284181\", \"tissue\"] = \"GMP 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284167\", \"tissue\"] = \"GMP 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284116\", \"tissue\"] = \"CMP 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284065\", \"tissue\"] = \"CMP 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284002\", \"tissue\"] = \"MEP 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284129\", \"tissue\"] = \"MEP 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284133\", \"tissue\"] = \"hematopoietic stem 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284013\", \"tissue\"] = \"hematopoietic stem 100 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284141\", \"tissue\"] = \"Purkinje 50 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284062\", \"tissue\"] = \"Purkinje 50 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284113\", \"tissue\"] = \"cerebral cortex, layer 5 50 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284164\", \"tissue\"] = \"cerebral cortex, layer 5 50 cells\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04284179\", \"tissue\"] = \"CH12.LX immortalized cell line\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04283998\", \"tissue\"] = \"CH12.LX immortalized cell line\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN04283997\", \"tissue\"] = \"CH12.LX immortalized cell line\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689605\", \"tissue\"] = \"thymus\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689601\", \"tissue\"] = \"ovary\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689599\", \"tissue\"] = \"lung\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689597\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689596\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689594\", \"tissue\"] = \"blood\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689604\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689603\", \"tissue\"] = \"spleen\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689602\", \"tissue\"] = \"skin\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689600\", \"tissue\"] = \"muscle\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689598\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2689595\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01853600\", \"tissue\"] = \"skin\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01853601\", \"tissue\"] = \"skin\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01853599\", \"tissue\"] = \"skin\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01853598\", \"tissue\"] = \"skin\"    \n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998531\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998528\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998533\", \"tissue\"] = \"skeletal muscle\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998535\", \"tissue\"] = \"ovary\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998532\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998529\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998536\", \"tissue\"] = \"blood\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998534\", \"tissue\"] = \"lung\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN00998537\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMD00004034\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMD00004035\", \"tissue\"] = \"skeletal muscle\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02781105\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02781095\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02781093\", \"tissue\"] = \"kidney\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02781090\", \"tissue\"] = \"brain\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02781081\", \"tissue\"] = \"heart\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02781078\", \"tissue\"] = \"skeletal muscle\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02781061\", \"tissue\"] = \"skin\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMD00043263\", \"tissue\"] = \"liver\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN01780059\", \"tissue\"] = \"testis\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMEA2553415\", \"tissue\"] = \"fibroblast\"\n",
    "    sra_table.loc[sra_table[\"biosample\"]==\"SAMN02205043\", \"tissue\"] = \"liver\"\n",
    "    \n",
    "    return sra_table\n",
    "\n",
    "\n",
    "def exclude_treatment_terms(sra_table, control_term_list):\n",
    "    print('Running: exclude_treatment_terms')\n",
    "    for item in control_term_list.items():\n",
    "        column = item[0]\n",
    "        sra_table[column] = sra_table[column].astype(str)\n",
    "        for control_term in item[1]:\n",
    "            if not control_term=='':\n",
    "                bioprojects = sra_table['bioproject'][sra_table[column].str.contains(control_term)].unique()\n",
    "                for bioproject in bioprojects:\n",
    "                    if bioproject != '':\n",
    "                        is_bioproject = (sra_table['bioproject']==bioproject)\n",
    "                        is_control_term = (sra_table[column].str.contains(control_term))\n",
    "                        sra_table.loc[(is_bioproject & -is_control_term), 'exclusion'] = 'treatment'\n",
    "    return sra_table\n",
    "\n",
    "def nspot_cutoff(sra_table, min_nspots=5000000):\n",
    "    print('Running: nspot_cutoff')\n",
    "    sra_table['total_spots'] = sra_table['total_spots'].fillna(0).astype(int)\n",
    "    sra_table.loc[-(sra_table['total_spots']==0) & (sra_table['total_spots'] < min_nspots), 'exclusion'] = 'low_nspots'\n",
    "    return sra_table\n",
    "\n",
    "def exclude_entry_with_keyword(sra_table):\n",
    "    print(\"Running: exclude_entry_with_keyword\")\n",
    "    sra_table.loc[~sra_table['antibody'].isnull(), 'exclusion'] = 'immunoprecipitation'   \n",
    "    sra_table.loc[~sra_table['cell'].isnull(), 'exclusion'] = 'cell_culture'\n",
    "    for col in ['exp_title','study_title','design','sample_title','sample_description','lib_name','protocol',]:\n",
    "        sra_table.loc[:,col] = sra_table.loc[:,col].astype(str)\n",
    "        sra_table.loc[sra_table[col].str.contains('RipSeq').fillna(False), 'exclusion'] = 'immunoprecipitation'\n",
    "        sra_table.loc[sra_table[col].str.contains('chrom_RNAseq').fillna(False), 'exclusion'] = 'immunoprecipitation'\n",
    "        sra_table.loc[sra_table[col].str.contains('RNAi').fillna(False), 'exclusion'] = 'RNAi'\n",
    "        sra_table.loc[sra_table[col].str.contains('shRNA RNA-seq').fillna(False), 'exclusion'] = 'small_RNA'\n",
    "        sra_table.loc[sra_table[col].str.contains('smRNA').fillna(False), 'exclusion'] = 'small_RNA'\n",
    "        sra_table.loc[sra_table[col].str.contains('CAGE').fillna(False), 'exclusion'] = 'CAGE'\n",
    "        sra_table.loc[sra_table[col].str.contains('piRNA').fillna(False), 'exclusion'] = 'small_RNA'\n",
    "        sra_table.loc[sra_table[col].str.contains('^exposed to').fillna(False), 'exclusion'] = 'treatment'\n",
    "    for col in ['treatment','protocol',]:\n",
    "        sra_table.loc[sra_table[col].str.contains('miRNA').fillna(False), 'exclusion'] = 'small_RNA'\n",
    "        sra_table.loc[sra_table[col].str.contains('sRNA').fillna(False), 'exclusion'] = 'small_RNA'\n",
    "        sra_table.loc[sra_table[col].str.contains('uranium').fillna(False), 'exclusion'] = 'treatment'\n",
    "    for col in ['age',]:\n",
    "        sra_table.loc[sra_table[col].str.contains('embryo').fillna(False), 'exclusion'] = 'embryonic'\n",
    "        sra_table.loc[sra_table[col].str.contains('Embryo').fillna(False), 'exclusion'] = 'embryonic'\n",
    "        sra_table.loc[sra_table[col].str.contains('somite').fillna(False), 'exclusion'] = 'embryonic'\n",
    "        sra_table.loc[sra_table[col].str.contains('cell').fillna(False), 'exclusion'] = 'embryonic'\n",
    "        sra_table.loc[sra_table[col].str.contains('Fertiliz').fillna(False), 'exclusion'] = 'embryonic'\n",
    "        sra_table.loc[sra_table[col].str.contains('fertiliz').fillna(False), 'exclusion'] = 'embryonic'\n",
    "        sra_table.loc[sra_table[col].str.contains('oocyte').fillna(False), 'exclusion'] = 'embryonic'\n",
    "    for col in ['genotype',]:\n",
    "        sra_table.loc[sra_table[col].str.contains('foxn3').fillna(False), 'exclusion'] = 'transgenic'\n",
    "    for col in ['exp_title','study_title','design','sample_title','sample_description','lib_name','experiment','treatment','protocol','age',]:\n",
    "        sra_table.loc[sra_table[col].str.contains('single cell').fillna(False), 'exclusion'] = 'single_cell'\n",
    "        sra_table.loc[sra_table[col].str.contains('Single cell').fillna(False), 'exclusion'] = 'single_cell'\n",
    "        sra_table.loc[sra_table[col].str.contains('Single Cell').fillna(False), 'exclusion'] = 'single_cell'\n",
    "    return sra_table\n",
    "\n",
    "def remove_redundant_biosample(sra_table):\n",
    "    print('Running: remove_redundant_biosample')\n",
    "    redundant_bool = sra_table.duplicated(subset=['bioproject', 'biosample'], keep = 'first')\n",
    "    sra_table.loc[redundant_bool, 'exclusion'] = 'redundant_biosample'\n",
    "    return sra_table\n",
    "\n",
    "def maximize_bioproject_sampling(df, target_n=10):\n",
    "    df['bioproject'] = df['bioproject'].fillna('unknown')\n",
    "    df['is_sampled'] = 'No'\n",
    "    df['is_qualified'] = 'No'\n",
    "    df.loc[(df.exclusion=='no'), 'is_qualified'] = 'Yes'\n",
    "    while len(df.loc[(df['is_sampled']=='Yes')&(df.exclusion=='no'),:]) < target_n:\n",
    "        if len(df) <= target_n:\n",
    "            df.loc[(df.exclusion=='no'), 'is_sampled'] = 'Yes'\n",
    "            break\n",
    "        else:\n",
    "            df_unselected = df.loc[(df['is_sampled']=='No')&(df.exclusion=='no'),:]\n",
    "            bioprojects = df_unselected['bioproject'].unique()\n",
    "            if len(bioprojects) == 0:\n",
    "                break\n",
    "            remaining_n = target_n - len(df.loc[df['is_sampled']=='Yes',:])            \n",
    "            select_n = min([len(bioprojects), remaining_n])\n",
    "            selected_bioprojects = numpy.random.choice(bioprojects, size=select_n, replace=False)\n",
    "            selected_index = []\n",
    "            for bioproject in selected_bioprojects:\n",
    "                index = numpy.random.choice(df_unselected.index[df_unselected['bioproject']==bioproject], size=1, replace=False)\n",
    "                selected_index.append(int(index))\n",
    "            df.loc[selected_index, 'is_sampled'] = 'Yes'\n",
    "    return df\n",
    "\n",
    "def label_sampled_data(sra_table, suppress_to=10, tissue_selection=[]):\n",
    "    print('Running: label_sampled_data')\n",
    "    sra_table_labeled = pandas.DataFrame()\n",
    "    species = sra_table['scientific_name'].unique()\n",
    "    for sp in species:\n",
    "        sp_table = sra_table.loc[sra_table['scientific_name'] == sp, :]\n",
    "        tissues = sp_table['tissue'].unique()\n",
    "        for tissue in tissues:\n",
    "            sp_tissue = sp_table.loc[sp_table['tissue'] == tissue, :]\n",
    "            sp_tissue = maximize_bioproject_sampling(df=sp_tissue, target_n=suppress_to)\n",
    "            sra_table_labeled = pandas.concat([sra_table_labeled, sp_tissue], axis=0)\n",
    "    sra_table_labeled['is_selected_tissue'] = 'Yes'\n",
    "    if len(tissue_selection):\n",
    "        sra_table_labeled.loc[-(sra_table_labeled['tissue'].isin(tissue_selection)),'is_selected_tissue'] = 'No'\n",
    "    return sra_table_labeled\n",
    "\n",
    "def make_cross_tables(sra_table, n_sp_cutoff=5, suppress_tissue=True, sample_upper=True):\n",
    "    print('Running: make_cross_tables')\n",
    "    if suppress_tissue:\n",
    "        sra_table = sra_table.loc[sra_table['is_selected_tissue']=='Yes',:]\n",
    "    if sample_upper:\n",
    "        sra_table = sra_table.loc[sra_table['is_sampled']=='Yes',:]    \n",
    "    sra_reduced = sra_table[['scientific_name', 'biosample', 'tissue']]\n",
    "    sra_pivot = sra_reduced.pivot_table(columns='tissue',index='scientific_name', aggfunc='count')\n",
    "    sra_pivot.columns = sra_pivot.columns.get_level_values(1)\n",
    "    column_sort = sra_pivot.count(axis='index').sort_values(ascending=False).index\n",
    "    index_sort = sra_pivot.count(axis='columns').sort_values(ascending=False).index\n",
    "    sra_pivot = sra_pivot.loc[index_sort,column_sort]\n",
    "    sra_pivot_reduced = sra_pivot.loc[:,sra_pivot.count(axis='index') >= n_sp_cutoff]\n",
    "    column_sort = sra_pivot_reduced.count(axis='index').sort_values(ascending=False).index\n",
    "    index_sort = sra_pivot_reduced.count(axis='columns').sort_values(ascending=False).index\n",
    "    sra_pivot_reduced = sra_pivot_reduced.loc[index_sort,column_sort]\n",
    "    return sra_pivot_reduced\n",
    "\n",
    "def reorder_columns(sra_table, omit_misc):\n",
    "    column_names = ['scientific_name','tissue','tissue_original','genotype','sex','age','treatment','source_name','is_sampled','is_selected_tissue',\n",
    "                    'is_qualified','exclusion',\n",
    "                    'protocol','bioproject','biosample','experiment',\n",
    "                    'run','sra_primary','sra_sample','sra_study','study_title','exp_title','design','sample_title','sample_description','lib_name',\n",
    "                    'lib_layout','lib_strategy','lib_source','lib_selection','instrument','total_spots','total_bases','size','nominal_length','nominal_sdev',\n",
    "                    'spot_length','read_index','read_class','read_type','base_coord','lab','center','submitter_id','pubmed_id',\n",
    "                    'taxid','published_date','biomaterial_provider','cell','location','antibody','batch','misc']\n",
    "    if omit_misc:\n",
    "        sra_table_reordered = sra_table.loc[:,column_names]\n",
    "    else:\n",
    "        misc_columns = [ col for col in sra_table.columns if col not in column_names ]\n",
    "        sra_table_reordered = sra_table.loc[:,column_names+misc_columns]\n",
    "    return sra_table_reordered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Curation lists\n",
    "column_aggregation_list = {\n",
    "    'age':[\"arrayexpress-developmentalstage\",\"agedays\",\"age_classification\",\"age_description\",\"age_in_years\",\"agecat\",\"dev_stage\",'ages',\n",
    "           \"development_satge\",\"development_stage\",\"developmental_stage\",\"developmentalstage\",\"differentiation_stages\",\"embryonic_day\",\n",
    "           \"embryonic_stage\",\"female_age\",\"sample_comment\",\"stage\",\"age_at_collection\",\"age_classification\",'age_description','age_in_years',\n",
    "           'age_of_collection','age_of_fly_in_days_post_eclosion','agecat','agedays','age{category}','animal_age_at_collection','day',\n",
    "           'arrayexpress-timepoint','avg_age_after_second_6_months_of_open_access','collection_timing_in_reference_to_e_lineage',\n",
    "           'day_relative_to_weaning','dev-stage','developmental_time','duration','embryo_stage','female_birth_date','gestational_age',\n",
    "           'gestational_age_at_sample_collection','hotta_stage','male_age','male_birth_date','male_collection_date','male_death_date',\n",
    "           'timepoint','trimester','birth_date','collection_date','death_date','female_collection_date','female_death_date','geo_loc_name',\n",
    "           'sample_date','sampling_date','specimen_collection_date','time_-_blastomeres','animal_length','date','developemental_stage',\n",
    "           'colection_date','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','',],\n",
    "    'antibody':['chip_antibody','chip_or_ip_antibody','clip-antibody','clip_antibody','rip_antibody','rna_binding_protein',\n",
    "                '','','','','','','','','','','','','','','','','',],\n",
    "    'batch':['adapter_barcode','animal','animal_id','animalid','barcode','biological_replicate','biorep','brain_code','clutch','cohort',\n",
    "            'collected_by','cow','custom_field','donator_monkey_number','donor_id','horse_number','iclip_barcode','id','illumina_rna-seq_bar_codes',\n",
    "            'index','individual','individual_id','individuals','internal_id','labanimalnumber','labexpid','mahpic_non_human_primate_individual_id',\n",
    "            'monkey_id','non_human_primate_individual_id','number','oocyte/embryo_number','pig_id','replicate','rna_processing_batch','run_id',\n",
    "            'sample','sample-type','sample_coding','sample_descriptor','sample_id','sample_identifier','sample_number','series','specimen_voucher',\n",
    "            'study','study_group','study_phase','subject','subject_id','subjectid','technical_replicate','tissue_abbreviation','tissue_code',\n",
    "            'umc_id','unique_id','uniqueid','wur_marker_type','animal_number','ebi_equivalent_biosample','ercc','flow_cell','flowcell','lane',\n",
    "             'lane_num','num_replicons','cell_barcode','submission_identifier','library_id','library_index_sequence_used_to_demultiplex',\n",
    "             'replicate_/_experiment','sequencing_pool','plate_col_id','plate_row_id','replicate_no.','run_no.','sire','unique_identifier',\n",
    "             '','','','','','','','','','','','','','','','','','','','','','','','','','','','','',],\n",
    "    'biomaterial_provider':[\"biosourceprovider\",\"female_biomaterial_provider\",\"male_biomaterial_provider\",'library_preparation_location',\n",
    "                    'material_provider','availability','','','',],\n",
    "    'bioproject':['bioproject_id','bioprojectid','','','','','',],\n",
    "    'cell':[\"cell_type\",\"cell_line\",\"cell_class\",\"cell_description\",\"cell_subtype\",\"arrayexpress-celltype\",\"culture_collection\",\"culture_conditions\",\n",
    "           'cell-line','cell-type','cell_or_tissue_type','eosinophil','feature','germ_layer','cells_derived_from','number_cells','cell_organism',\n",
    "            'cell_culture_protocol','cell_population','cell_typing','','','','','','','','','',],\n",
    "    'protocol':[\"extract_protocol\",\"female_sample_collection_protocol\",\"protocol\",\"sample_extraction_method\",'experimental_protocol',\n",
    "                  \"chemical_treatment_of_isolated_rna\",'amplification','assay','experiment_target','extraction_protocol','fixation','fraction',\n",
    "                  'library_selection','library_source','library_strategy','library_type','libraryprotocol','lysis_buffer_ion_concentrations',\n",
    "                  'lysis_strategy','male_sample_collection_protocol','meoh-fixed','molecule_subtype','prep_type','preperation_kit',\n",
    "                  'purification_protocol','rin','rna_fraction','rna_rin_values_after_globin_depletion','rna_rin_values_before_globin_depletion',\n",
    "                  'rna_subtype','sample_material','sample_storage','sample_storage_processing','small_rna_classes','specimen_collection_protocol',\n",
    "                  'specimen_with_known_storage_state','tissue_state','store_cond','datatype_description','datatype','monosome_enrichment_strategy',\n",
    "                  'rna_concentration_after_globin_depletion','rna_concentration_before_globin_depletion','rna_input','sequencing_type','date_run',\n",
    "                  'instrument_model','ngs_platform','paired_end_seq?','paired_or_single-end','pe_read_length','platform','quality','route',\n",
    "                  'read_len_orig','readtype','readtype_description','sequencer','sequencing_method','extraction','indrops_version',\n",
    "                'minimum_counts_per_cell_threshold_used_to_remove_background_barcodes','minimum_reads_per_cell_during_initial_processing',\n",
    "                'spike-in','','','','','','','','','','','','','','','','','','',],\n",
    "    'nominal_length':['mean_insert_size','','','','','','','','','','','','','','','','','','',],\n",
    "    'genotype':[\"arrayexpress-genotype\",\"arrayexpress-phenotype\",\"arrayexpress-rnai\",\"background_strain\",\"breed\",\"cultivar\",\"chicken_line\",\n",
    "                \"ecotype\",\"full_genotype\",\"genetic_background\",\"genetic_line\",\"genetic_modification\",\"genotype/variaion\",\"genotype/variation\",\n",
    "                \"germline_knock-down\",\"germline_knockdown_and_other_transgenes\",\"isolate\",\"line\",\"phenotype\",\"ploidy\",\"population\",\n",
    "                \"snp_allelic_state\",\"strain\",\"strain/background\",\"strain_origin\",\"strains\",\"sub_species\",\"subspecies\",\"variety\",'breed_name',\n",
    "                'arrayexpress-strainorline','breed/line','breed_type','breeding_history','breeds','chicken_breed','clonality','marker',\n",
    "                'dstim_knockdown_status','environmental_history','fly_line','origin','phenotype/variation','pig_breed','pig_type','propagation',\n",
    "                'reporter_construct','rhesus_monkey_origin','rna_interference','strain_background','strain_description','strain_prrsv','strainorline',\n",
    "                'transfecting_vector','transgenic_strain','virus_group','arrayexpress-species','parental_strain','tax_id','transgene',\n",
    "                'genotype/phenotype','cross','dad','driver','qtl_genotype/haplotype','','','','','','','','','','','','','','','','','','','','','','','','',],\n",
    "    'location':['birth_location','country','feeding_barn','female_birth_location','geographic_location','lat_lon','library_preparation_location_latitude',\n",
    "              'library_preparation_location_latitude_units','library_preparation_location_longitude','library_preparation_location_longitude_units',\n",
    "              'local_origin','male_birth_location','sampling_site','seq_center','sequencing_location','sequencing_location_latitude','station',\n",
    "              'sequencing_location_latitude_units','sequencing_location_longitude','sequencing_location_longitude_units','source','biome',\n",
    "                'env_biome','env_feature','env_material','environment','environmental_package','','','','','','','','','','','','','','','','','','','','','','','','','','','','',],\n",
    "    'misc':['ena-checklist','ena-first-public','ena-last-update','estimated_size','investigation_type','lab_description','library_layout','mapalgorithm',\n",
    "            'mapalgorithm_description','new_attribute','notes','organism','project','project_name','species','sra_experiment_accession',\n",
    "            'biosamplemodel','sample_year','number_of_pieces','specimen_picture_url','submission_description','submission_title','unknown',\n",
    "            'body_weight','file_no.','num_parts_in_pool','processing_date','','','','','','','','','','','',],\n",
    "    'nominal_length':['mean_insert_size','','','','','','','','','','','','','','','',],\n",
    "    'sex':[\"arrayexpress-sex\",\"gender\",\"host_sex\",'cell_sex','gender_type','obsolete_sex','both_sexes_pooled','','',],\n",
    "    'tissue':[\"arrayexpress-organismpart\",\"bio_material\",\"body_site\",\"biopsy_site\",\"brain_region\",\"description\",\"experiment_set\",\"explant\",\n",
    "              \"isolation-source\",\"isolation_source\",\"label\",\"mixedtissues\",\"muscle_type\",\"organ/tissue\",\"organ_part\",\"organism_part\",\n",
    "              \"organismpart\",\"oviduct\",\"region\",\"sample_name\",\"sample_origin\",\"sampling_position\",\"source_name\",\"tag\",\"tissue-type\",\n",
    "              \"tissue/cell\",\"tissue/cell_type\",\"tissue_location\",\"tissue_or_dev_stage\",\"tissue_origin\",\"tissue_source\",\"tissue_type\",\"title\",\n",
    "              \"arrayexpress-organismpart\",\"organism_part\",'embryo_region','organsim_part','tissue_/_cells','tissue_lib','tissue_region',\n",
    "              '','','','','','','','','','','','','','','','','','','',],\n",
    "    'treatment':[\"agent\",\"arrayexpress-diseasestate\",\"arrayexpress-growthcondition\",\"arrayexpress-immunoprecipitate\",\"behavior\",\n",
    "                 \"biopsy_day\",\"challenged_with\",\"chemical_treatment_of_isolated_rna\",\"diet\",\"diet_fed\",\"dietary_group\",\"disease\",\"disease_status\",\n",
    "                 \"domestication\",\"drip\",\"energy_balance\",\"fed_status\",\"feeding\",\"feeding_type\",\"fertility_group\",\"food\",\"group\",\"health_state\",\n",
    "                 \"infect\",\"infected_with\",\"infection\",\"infection_status\",\"inoculation\",\"maternal_diet\",\"meat_quality\",\"mptp_treatment\",\"oxidation\",\n",
    "                 \"pathogen\",\"percent_marinade_uptake_at_24h_post-slaughter\",\"phenotype_sample_type\",\"restriction_model\",'altitude','dpi',\n",
    "                 \"salmonella_shedding_status\",\"status\",\"stimulation\",\"stimulus\",\"stress\",\"survival\",\"transplant_type\",\"treated_with\",\n",
    "                 \"treatment_group\",\"pregnancy_outcome\",\"isol_growth_condt\",'agent','aliquot','ammonia_concentration','ammonia_exposure',\n",
    "                 'average_altitude','blastocyst_rate','breeding_method','bull_ntm','calcium_intake','cell_status','cellular_compartment',\n",
    "                 'chemical','collection_time','comment','concentration','condition','culture_condition','culture-collection','culture_type',\n",
    "                 'custom_name','day_of_infection','day_post_infection','days_at_29_c','days_post_infection','differentiation_time','disease_stage',\n",
    "                 'drinking_category','drug','drug_for_synchronization','egg_production','enzyme_treatment','exposure_time','extracellular_component',\n",
    "                 'fasted_status','fed_for','feed_efficiency','feeding_period','feeding_treatment','growth_condition','growth_protocol',\n",
    "                 'harmonized_disease_score','harvest_time','health-state','health_status','health_status_at_collection','heat_hours',\n",
    "                 'hematocrit','hemoglobin_measurement','host','host_tissue_sampled','hours_at_restrictive_temperature','hp-prrsv_infection',\n",
    "                 'infection_route','infectious_agent','inflammatory_stimulus_treatment','injury','kinetics','lens_type','lesion_score','lesion_number',\n",
    "                 'library_name','litter','loin_weight_group','lps_exposure','material','maternal_treatment','mating_status','mock','mode','morpholino',\n",
    "                 'muscle_name','neutralization_group','normalized?','number_born_alive','oxygen_concentration','parity_number','passage',\n",
    "                 'passage_number','passagers','passages','phase','pooled','pregnancy_status','pregnancy_time','purity','resorted',\n",
    "                 'rna_treatment','sample_group','sample_processing','sampleid','sampling_time_point','sampling_to_preparation_interval',\n",
    "                 'selection_criteria','serologic_response_status','shear_force','skin_color','social_rank','sorted','specific_host','stim_time',\n",
    "                 'stud_book_number','superovulated','temperature','thermal_treatment','time','time-point','time_point','ega-disease',\n",
    "                 'time_post_infection','timepoint_postchallenge','training','training_timepoint','treatment_description','tumor','tumor_grading',\n",
    "                 'vaccine','vaccine_administration','vaccine_group','viremia_level','virus','virus_dose','d','days_post-infection','ega-subject',\n",
    "                 'ega-treatment','ercc_pool','host_taxid','lymphocyte','monocyte','morphology','neutrophil','platelet_count','rfi_value',\n",
    "                 'sample_type','source_of_female','sperm_dna_fragmentation_index','srek_ligation','time/hpf','total_number_born','treatments',\n",
    "                 'type','weight','arrayexpress-compound','arrayexpress-dose','assay_type','backfat_thickness_of_live(mm)','basophil',\n",
    "                 'photorecptors_remaining','productivity','hours/days_after_injury','disease_state','mehg_exposure','operation',\n",
    "                 'post-injury_time_point','embryo_source','earlobe_color','reproductive_status','specimen_size','specimen_volume','specimen_weight',\n",
    "                 'liver_phenotype','clinical_information','oestrous_peroid','secondary_description','finishing_diet','gestation_duration',\n",
    "                 'growth_rate','heart_weight','information','light_treatment','maternal_food','number_of_passages','paternal_experience',\n",
    "                 'personal_experience','physiological_conditions','rfi_group','sample_role','slaughter_weight','target','target_source',\n",
    "                 'temperature_regimen','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','','',],\n",
    "}\n",
    "\n",
    "tissue_aggregation_list = {\n",
    "    'adrenal':['adrenal gland',],\n",
    "    'adipose':[ \"white adipose\",\"adipose   subcutaneous\", \"adipose   visceral\", \"subcutaneous adipose\", \"fat\",'white adipose subcutaneous',\n",
    "    'cardiac adipose','adipose tissue','brown adipose tissue','cardiac adipose','subcutaneous adipose tissue','','','','','','','','','','','','','','',],\n",
    "    'bone':['bones',],\n",
    "    'blood':[\"whole blood\",\"blood, peripheral vein\",\"blood sample\",],\n",
    "    'brain':[ \"a brain\", \"b brain\", \"normal brain\", \"whole brain\", \"brain, whole\", \"whole brains\", \"brain cerebrum\", \"brain left hemisphere\", \n",
    "             \"brain right hemisphere\", \"male  brain\", \"female brain\", \"brain ground control\", \"larval brain\", \"brain tissue\",'brain gl',\n",
    "             'brain without ventral nerve cord','brain space flight',],\n",
    "    'breast':[\"breast   mammary tissue\",],\n",
    "    'cerebellum':[\"brain cerebellum\",\"brain, cerebellum\",\"brain   cerebellum\",'cerebellar hemisphere','','','','','','','','','','','','','','',],\n",
    "    'cerebral cortex':[\"dog cerebral cortex\",],\n",
    "    'colon':['ascending colon','descending colon','gut colon','spiral colon','sigmoid colon','','','','','','','','','','','','','','','',],\n",
    "    'embryo':[ \"embryos\", \"single embryo\", \"single embryo \", \"whole embryo\", \"whole embryos\", \"whole embryo of  individuals\", \"whole-embryo\", \n",
    "              \"whole embryo homogenate\", \"embryo, whole, . days gestation  (n=)\", \"day  elongated embryo pool\", \"day  hatched embryo pool\", \n",
    "              \"body tissue\", \"mixed sex embryos\",'about  h old embryos','mrnaseq of whole embryo','ivf embryos','living one cell embryo',\n",
    "              'older embryos','post gastrulation embryo','two to four cell embryo','wild type embryos, tap treated','whole embryo, min intervals',\n",
    "              'cell embryo','mixed stage embryos','early embryo','late gastrula embryo','dissociated embryo','whole embryo stage .',\n",
    "              'embryos,  hr after egg laying','hr embryo','whole   hour embryo','total embryo','mrnaseq of whole embryo','whole embryo lysate',\n",
    "              'intact embryo','hh whole embryo','cell embryo at pronuclear','blastocyst stage embryo','morula stage embryo','eight cell stage embryo',\n",
    "              'four cell stage embryo','two cell stage embryo','pronuclear stage embryo','embryo, whole, . days gestation','embryo whole embryo',\n",
    "              'hpf whole embryo','drosophila embryo','single stage  embryo','','','','','','','','','','','','','','','','','','','','','','','','','','','','','',],\n",
    "    'fibroblast':[ \"skin fibroblasts from adult naked mole-rat.\", \"skin fibroblasts from adult naked mole rat.\", \"fibroblast, skin\", \n",
    "                  \"primary skin fibroblasts\", \"skin fibroblast\",'','','','','','','','','','','','','','','','','','',],\n",
    "    'frontal cortex':[\"brain frontal cortex\", \"brain, frontal cortex\", \"brain   frontal cortex\",'','','','','','','','','','','','','','','','','','',], \n",
    "    'gill':['gills','','','','','','','','','','','','','','','','','','',],\n",
    "    'heart':[\"a heart\", \"b heart\", \"male  heart\", \"female heart\", \"the total rnas were isolated from the chimpanzee heart with trizol\",\n",
    "             'heart/coronary','heart, left ventricle','heart left ventricle','','','','','','','','','','','','','','','','','',],\n",
    "    'hippocampus':['brain   hippocampus','hippocampus ca','','','','','','','','','','','','','','','','','','',],\n",
    "    'hypothalamus':[\"dog hypothalamus\",\"brain   hypothalamus\",'','','','','','','','','','','','','','','','','','',],\n",
    "    'kidney':[ \"normal kidney\", \"a left kidney\", \"a right kidney\", \"male  kidney\", \"female kidney\", \"head kidney\",\"head kidney tissue\",],\n",
    "    'liver':[\"normal liver\", \"a liver\", \"b liver\", \"male  liver\", \"female liver\", \"liver ground control\", \"adult liver\",'liver space flight','newborn liver',\n",
    "             'liver left lateral lobe','','','','','','','','','','','','','','','','','',],\n",
    "    'lung':[ \"left lung\", \"lung healthy\", \"lung control \", \"lung control\", \"whole lung\", \"a lung\", \"Lung control_\", \"Lung control_1\", \n",
    "            \"Lung control_2\", \"Lung control_3\", \"male  lung\", \"female lung\", \"lung/trachea\",'lung day','lungs','','','','','','','','','','','','','','','','','','',],\n",
    "    'lymph node':[ \"Ferret, lymph nodes, control\", \"normal lymph node\", \"lymph nodes\", \"supramammary lymph node\", \n",
    "                  \"submandibular lymph node\", \"popliteal lymph node\", \"mesenteric lymph node\",'bronchial lymph node',\n",
    "                  'abomasal lymph node','ferret, lymph nodes, control','intestinal lymph node','lymph node mesenteric',\n",
    "                  'lymph node prescapular','para rectal lymph node','prescapular lymph node','tracheobronchial lymph node',\n",
    "                  'tracheobronchial lymph node, iav s     dpi siv ecu','mesenteric lymphonodes','axillary lymphonodes','','','','','','','',],\n",
    "    'mushroom body':['mshroom body',],\n",
    "    'oocyte':['mature oocyte','stage  oocyte','','','','','','','','','','','','','','','','','',],\n",
    "    'ovary':[ \"ovaries\", \"virgin female ovary\", \"adult ovary\", \"adult ovaries\", \"adult drosophila ovaries\", \"ovary ground control\", \"left ovary\", \n",
    "             \"right ovary\",'adult female ovary, no treatment, no gal activation','overy','ovary germarium to stage','ovary stage  + egg chambers',\n",
    "             'ovary stage  to','ovarian tissues','ovary space flight','ovary tissue','','','','','','','','','','','',],\n",
    "    'pancreas':['female pancreas','female  pancreas','','','','','','','','','','','','','','','','','','',],\n",
    "    'placenta':['placent','placenta + membranes','term placenta','','','','','','','','','','','','','','','','','',],\n",
    "    'pituitary':['brain pituitary','pituitary gland','anterior pituitary','','','','','','','','','','','','','','','','','',],\n",
    "    'prefrontal cortex':[\"brain, prefrontal cortex\",],\n",
    "    'retina':[ \"retina, rpe, choroid\", \"central retina\", \"peripheral retina\", \"whole retina\", \"adult retina\",],\n",
    "    'salivary gland':['salivary glands','major salivary gland','right salivary gland','left salivary gland',],\n",
    "    'skeletal muscle':[ \"muscle\", \"female  skeletal muscle\", \"skeletal_muscle\", \"body muscle\", \"muscle tissues\", \"muscle   skeletal\", \n",
    "                       \"muscle biceps\", \"muscle long dorsal\", \"biceps brachii\", \"pelvic limb\", \"longissimus muscle\", \"longissimus lumborum muscle\", \n",
    "                       \"longissimus dorsi\", \"semimembranosus muscle\", \"breast musle\", \"breast muscle\", \"longissimus thoracis\", \"gluteus medius muscle\", \n",
    "                       \"pectoralis superficialis muscle\", \"pectoralis major muscle\", \"psoas major muscle\", \"longissimus dorsi muscle\", \n",
    "                       \"biceps femoris muscle\", \"fast muscle\", \"slow muscle\", \"indirect flight muscle\", \"longissimus lumborum\", \"body wall muscle\", \n",
    "                       \"pectoralis\", \"longuissimus dorsi\", \"skeletal muscle tissue of biceps brachii\", \"longissimus thoracis muscle\", \"longissimus muslce\", \n",
    "                       \"pectoralis major\", \"soleus muscle\", \"pectoral muscle\", \"gluteus medius\",'leg musle','leg muscle','longissimus muscles',\n",
    "                       'longissimus dorsi muscle tissues','skeletal muscle tissue','thoracic muscle','thorax skeletal muscle','infraspinatus muscle, pre',\n",
    "                       'longissimus dorsi muscle tissueslongissimus dorsi muscle tissues','muscle abdominal','muscle gastrocnemian','crico arytenoid muscle',\n",
    "                       'ventral lateral sacrocaudal muscle','gluteal muscle','','','','','','','',],\n",
    "    'skin':[ \"skin, upper arm\", \"skin, unspecified\", \"black skin\", \"white skin\",   \"back skin\", \"dorsal skin\", \"ventral skin\", \"skin back\", \"skin side\", \n",
    "            \"male skin\", \"female  skin\", \"skin   sun exposed\", \"skin   not sun exposed\", \"tarsometatarsal skin\", \"thigh skin\", \"ear skin\", \"stkin\",\n",
    "            'wing skin','earlobe skin','skin tissue of inner side of the wing','abdomen skin tissue','skin of back','','','','','','','','','','','','','','','',],\n",
    "    'sperm':['f sperm','','','','','','','','','','','','','','','','','','',],\n",
    "    'spinal cord':[\"brain   spinal cord\",],\n",
    "    'spleen':[ \"spleen tissue\", \"a spleen\", \"b spleen\", \"d spleen\", \"male spleen\", \"female  spleen\", \"chicken splenic tissue\",'pleen',],\n",
    "    'temporal lobe':[\"brain, temporal lobe\",\"brain temporal lobe\",\"temperal lobe\",],\n",
    "    'tongue':['tongue.',],\n",
    "    'testis':[ \"testes\", \"super bull testis\", \"adult testis\", \"male  testis\", \"virgin testes\", \"testis ground control\", \"dissected testis\", \"mature testis\", \n",
    "              \"adult testes\",'fly testis','testis space flight','whole testes','','','','','','','','','','','','','','','','',],\n",
    "    'whole organism':[\"whole worms\",\"whole fly\",'worm','whole body','whole worm','whole animal','whole flies','whole fish','','','','','','','','','','',],\n",
    "}\n",
    "\n",
    "control_term_list = {\n",
    "    'treatment':[\"mock\", \"control\", \"Control\", \"none\", \"CNTL\", \"CON\", \"normal\", \"untreated\", \"Mock\", \"mock\", \"Ad Libitum\", \"uninfected\", \n",
    "                 \"Uninfected\", \"in the absence of\", \"kontrol\", \"None\", \"Thermoneutral\",'21%[oxygen_concentration]','DMSO','Normal',' 0hrs',\n",
    "                 '^0 ppm','un-infected','','','','','',],\n",
    "    'genotype':[\"wild\", \"Wild\", \"N2\", \"OregonR\", \"Oregon R\", \"Canton S\", \"Canton-S\", \"y w\", \"yw\", \"wt\",'\\+\\/\\+'],\n",
    "    'sample_title':[\"wild\",\"Wild type\",\"control\",\"Control\",\"CON_\",],\n",
    "    'sample_description':[\"control\",],\n",
    "    'lib_name':[\"control\",],\n",
    "    'exp_title':[\"control\",\"Control_\",\"CON_\",\"TimePoint1_\",\"0 days post infection\",],\n",
    "}\n",
    "\n",
    "tissue_selection = [\"liver\",\"brain\",\"kidney\",\"testis\",\"heart\",\"ovary\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Entrez search conditions\n",
    "Entrez.email = \"YOUR_EMAIL_ADDRESS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species for SRA search: 105\n"
     ]
    }
   ],
   "source": [
    "# Read Ensembl species list\n",
    "ensembl_table = pandas.read_csv(\"/Users/kf/Dropbox/kfdata/02_Data/my_db/Ensembl/release-91/species_list/Ensembl_Release91_species_list.csv\", sep=\",\")\n",
    "ensembl_sp = ensembl_table[\"Scientific name\"].unique()\n",
    "ensembl_sp_binom = sorted(set([re.sub(r'(.+)(\\s)(.+)(\\s)(.+)', r\"\\1\\2\\3\", sp) for sp in ensembl_sp]))\n",
    "print('Number of species for SRA search:', len(ensembl_sp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ailuropoda melanoleuca\n",
      "Ailuropoda melanoleuca : reading xml from file\n",
      "Anas platyrhynchos\n",
      "Anas platyrhynchos : reading xml from file\n",
      "Anolis carolinensis\n",
      "Anolis carolinensis : reading xml from file\n",
      "Aotus nancymaae\n",
      "Aotus nancymaae : reading xml from file\n",
      "Astyanax mexicanus\n",
      "Astyanax mexicanus : reading xml from file\n",
      "Bos taurus\n",
      "Bos taurus : reading xml from file\n",
      "Caenorhabditis elegans\n",
      "Caenorhabditis elegans : reading xml from file\n",
      "Callithrix jacchus\n",
      "Callithrix jacchus : reading xml from file\n",
      "Canis lupus\n",
      "Canis lupus : reading xml from file\n",
      "Carlito syrichta\n",
      "Carlito syrichta : reading xml from file\n",
      "Cavia aperea\n",
      "Cavia aperea : reading xml from file\n",
      "Cavia porcellus\n",
      "Cavia porcellus : reading xml from file\n",
      "Cebus capucinus\n",
      "Cebus capucinus : reading xml from file\n",
      "Ceratotherium simum\n",
      "Ceratotherium simum : reading xml from file\n",
      "Cercocebus atys\n",
      "Cercocebus atys : reading xml from file\n",
      "Chinchilla lanigera\n",
      "Chinchilla lanigera : reading xml from file\n",
      "Chlorocebus sabaeus\n",
      "Chlorocebus sabaeus : reading xml from file\n",
      "Choloepus hoffmanni\n",
      "Choloepus hoffmanni : reading xml from file\n",
      "Chrysemys picta\n",
      "Chrysemys picta : reading xml from file\n",
      "Ciona intestinalis\n",
      "Ciona intestinalis : reading xml from file\n",
      "Ciona savignyi\n",
      "Ciona savignyi : reading xml from file\n",
      "Colobus angolensis\n",
      "Colobus angolensis : reading xml from file\n",
      "Cricetulus griseus\n",
      "Cricetulus griseus : reading xml from file\n",
      "Danio rerio\n",
      "Danio rerio : reading xml from file\n",
      "Dasypus novemcinctus\n",
      "Dasypus novemcinctus : reading xml from file\n",
      "Dipodomys ordii\n",
      "Dipodomys ordii : reading xml from file\n",
      "Drosophila melanogaster\n",
      "Drosophila melanogaster : reading xml from file\n",
      "Echinops telfairi\n",
      "Echinops telfairi : reading xml from file\n",
      "Equus caballus\n",
      "Equus caballus : reading xml from file\n",
      "Erinaceus europaeus\n",
      "Erinaceus europaeus : reading xml from file\n",
      "Felis catus\n",
      "Felis catus : reading xml from file\n",
      "Ficedula albicollis\n",
      "Ficedula albicollis : reading xml from file\n",
      "Fukomys damarensis\n",
      "Fukomys damarensis : reading xml from file\n",
      "Gadus morhua\n",
      "Gadus morhua : reading xml from file\n",
      "Gallus gallus\n",
      "Gallus gallus : reading xml from file\n",
      "Gasterosteus aculeatus\n",
      "Gasterosteus aculeatus : reading xml from file\n",
      "Gorilla gorilla\n",
      "Gorilla gorilla : reading xml from file\n",
      "Heterocephalus glaber\n",
      "Heterocephalus glaber : reading xml from file\n",
      "Homo sapiens\n",
      "Homo sapiens : reading xml from file\n",
      "Ictidomys tridecemlineatus\n",
      "Ictidomys tridecemlineatus : reading xml from file\n",
      "Jaculus jaculus\n",
      "Jaculus jaculus : reading xml from file\n",
      "Latimeria chalumnae\n",
      "Latimeria chalumnae : reading xml from file\n",
      "Lepisosteus oculatus\n",
      "Lepisosteus oculatus : reading xml from file\n",
      "Loxodonta africana\n",
      "Loxodonta africana : reading xml from file\n",
      "Macaca fascicularis\n",
      "Macaca fascicularis : reading xml from file\n",
      "Macaca mulatta\n",
      "Macaca mulatta : reading xml from file\n",
      "Macaca nemestrina\n",
      "Macaca nemestrina : reading xml from file\n",
      "Mandrillus leucophaeus\n",
      "Mandrillus leucophaeus : reading xml from file\n",
      "Meleagris gallopavo\n",
      "Meleagris gallopavo : reading xml from file\n",
      "Melopsittacus undulatus\n",
      "Melopsittacus undulatus : reading xml from file\n",
      "Mesocricetus auratus\n",
      "Mesocricetus auratus : reading xml from file\n",
      "Microcebus murinus\n",
      "Microcebus murinus : reading xml from file\n",
      "Microtus ochrogaster\n",
      "Microtus ochrogaster : reading xml from file\n",
      "Monodelphis domestica\n",
      "Monodelphis domestica : reading xml from file\n",
      "Mus caroli\n",
      "Mus caroli : reading xml from file\n",
      "Mus musculus\n",
      "Mus musculus : reading xml from file\n",
      "Mus pahari\n",
      "Mus pahari : reading xml from file\n",
      "Mus spretus\n",
      "Mus spretus : reading xml from file\n",
      "Mustela putorius\n",
      "Mustela putorius : reading xml from file\n",
      "Myotis lucifugus\n",
      "Myotis lucifugus : reading xml from file\n",
      "Nannospalax galili\n",
      "Nannospalax galili : reading xml from file\n",
      "Nomascus leucogenys\n",
      "Nomascus leucogenys : reading xml from file\n",
      "Notamacropus eugenii\n",
      "Notamacropus eugenii : reading xml from file\n",
      "Ochotona princeps\n",
      "Ochotona princeps : reading xml from file\n",
      "Octodon degus\n",
      "Octodon degus : reading xml from file\n",
      "Oreochromis niloticus\n",
      "Oreochromis niloticus : reading xml from file\n",
      "Ornithorhynchus anatinus\n",
      "Ornithorhynchus anatinus : reading xml from file\n",
      "Orycteropus afer\n",
      "Orycteropus afer : reading xml from file\n",
      "Oryctolagus cuniculus\n",
      "Oryctolagus cuniculus : reading xml from file\n",
      "Oryzias latipes\n",
      "Oryzias latipes : reading xml from file\n",
      "Otolemur garnettii\n",
      "Otolemur garnettii : reading xml from file\n",
      "Ovis aries\n",
      "Ovis aries : reading xml from file\n",
      "Pan paniscus\n",
      "Pan paniscus : reading xml from file\n",
      "Pan troglodytes\n",
      "Pan troglodytes : reading xml from file\n",
      "Papio anubis\n",
      "Papio anubis : reading xml from file\n",
      "Papio hamadryas\n",
      "Papio hamadryas : reading xml from file\n",
      "Pelodiscus sinensis\n",
      "Pelodiscus sinensis : reading xml from file\n",
      "Peromyscus maniculatus\n",
      "Peromyscus maniculatus : reading xml from file\n",
      "Petromyzon marinus\n",
      "Petromyzon marinus : reading xml from file\n",
      "Physter macrocephalus\n",
      "Physter macrocephalus : reading xml from file\n",
      "Poecilia formosa\n",
      "Poecilia formosa : reading xml from file\n",
      "Pongo abelii\n",
      "Pongo abelii : reading xml from file\n",
      "Procavia capensis\n",
      "Procavia capensis : reading xml from file\n",
      "Propithecus coquereli\n",
      "Propithecus coquereli : reading xml from file\n",
      "Pteropus vampyrus\n",
      "Pteropus vampyrus : reading xml from file\n",
      "Rattus norvegicus\n",
      "Rattus norvegicus : reading xml from file\n",
      "Rhinopithecus bieti\n",
      "Rhinopithecus bieti : reading xml from file\n",
      "Rhinopithecus roxellana\n",
      "Rhinopithecus roxellana : reading xml from file\n",
      "Saccharomyces cerevisiae\n",
      "Saimiri boliviensis\n",
      "Saimiri boliviensis : reading xml from file\n",
      "Sarcophilus harrisii\n",
      "Sarcophilus harrisii : reading xml from file\n",
      "Sorex araneus\n",
      "Sorex araneus : reading xml from file\n",
      "Sus scrofa\n",
      "processing  0 - 999\n",
      "processing  1000 - 1999\n",
      "processing  2000 - 2304\n",
      "Taeniopygia guttata\n",
      "Taeniopygia guttata : reading xml from file\n",
      "Takifugu rubripes\n",
      "Takifugu rubripes : reading xml from file\n",
      "Tetraodon nigroviridis\n",
      "Tetraodon nigroviridis : reading xml from file\n",
      "Tupaia belangeri\n",
      "Tupaia belangeri : reading xml from file\n",
      "Tursiops truncatus\n",
      "Tursiops truncatus : reading xml from file\n",
      "Vicugna pacos\n",
      "Vicugna pacos : reading xml from file\n",
      "Xenopus tropicalis\n",
      "Xenopus tropicalis : reading xml from file\n",
      "Xiphophorus maculatus\n",
      "Xiphophorus maculatus : reading xml from file\n"
     ]
    }
   ],
   "source": [
    "# Get SRA table\n",
    "sra_table_species = dict()\n",
    "for sp in ensembl_sp_binom:\n",
    "    print(sp)\n",
    "    sp_file_name = \"sra_table_sp_\"+sp.replace(' ', '_')+'_'+d+\".tsv\"\n",
    "    if sp == \"Homo sapiens\":\n",
    "        bioprojects = [\"PRJEB2445\",\"PRJNA143955\",\"PRJNA143627\",\"PRJNA154333\",\"PRJNA186646\",\"PRJNA272542\",\"PRJNA276463\",\n",
    "                       \"PRJNA233428\",\"PRJNA280600\"]\n",
    "    elif sp == \"Mus musculus\":\n",
    "        bioprojects = [\"PRJNA66167\",\"PRJNA176589\",\"PRJNA177791\",\"PRJNA184055\",\"PRJNA186646\",\"PRJNA324710\",\"PRJNA317431\",\n",
    "                       \"PRJNA315706\"]\n",
    "    elif sp == \"Danio rerio\":\n",
    "        bioprojects = [\"PRJNA276667\", \"PRJNA260259\", \"PRJEB2368\", \"PRJNA207719\", \"PRJNA255848\",\"PRJNA293388\",\"PRJNA379145\",\n",
    "                      \"PRJNA297460\",\"PRJNA317597\",\"PRJNA266803\",\"PRJNA263496\",\"PRJNA431371\",\"PRJNA371613\"]\n",
    "    elif sp == \"Rattus norvegicus\":\n",
    "        bioprojects = [\"PRJEB3197\",\"PRJNA177791\",\"PRJNA184055\",\"PRJNA218629\",\"PRJNA238328\",\"PRJEB6938\",\"PRJNA236460\",\n",
    "                       \"PRJNA264000\",\"PRJNA325290\"]\n",
    "    elif sp == \"Saccharomyces cerevisiae\":\n",
    "        continue\n",
    "    else:\n",
    "        bioprojects = []\n",
    "    if os.path.exists(sp_file_name):\n",
    "        print(sp, ': reading from tsv file.')\n",
    "        sra_table_species[sp] = pandas.read_csv(sp_file_name, sep='\\t', header=0, low_memory=False)\n",
    "    else:\n",
    "        search_term = get_search_term(species_name=sp, bioprojects=bioprojects, biosamples=[])\n",
    "        root = fetch_sra_xml(species_name=sp, search_term=search_term, save_xml=False, read_from_existing_file=False)\n",
    "        sra_table_species[sp] = sra_xml2table(species_name=sp, xml_root=root)\n",
    "        sra_table_species[sp].to_csv(sp_file_name, sep=\"\\t\", index=False)\n",
    "sra_table_multisp = pandas.concat(sra_table_species.values())\n",
    "sra_table_multisp.to_csv(\"sra_table_multisp_\"+d+\".tsv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GTEx project with >10k samples\n",
    "do_gtex=False\n",
    "if do_gtex:\n",
    "    sp = \"Homo sapiens\"\n",
    "    bioprojects = [\"PRJNA75899\"]\n",
    "    search_term = get_search_term(species_name=sp, bioprojects=bioprojects, biosamples=[])\n",
    "    root = fetch_sra_xml(species_name=sp, search_term=search_term, save_xml=True)\n",
    "    sra_table_gtex = sra_xml2table(species_name=sp, xml_root=root)    \n",
    "    sra_table_gtex.to_csv(\"sra_table_GTEx_\"+d+\".tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing  0 - 100\n",
      "processing  0 - 9\n"
     ]
    }
   ],
   "source": [
    "# Additional Bioprojects and Biosamples\n",
    "sra_table_additional = pandas.DataFrame()\n",
    "species_bioprojects = [\"Sus scrofa\", \"PRJEB1213\",]\n",
    "species = species_bioprojects[0::2]\n",
    "bioprojects = species_bioprojects[1::2]\n",
    "for (sp,bioproject) in zip(species,bioprojects):\n",
    "    search_term = bioproject\n",
    "    root = fetch_sra_xml(species_name=sp, search_term=search_term, save_xml=True)\n",
    "    sra_table_tmp = sra_xml2table(species_name=sp, xml_root=root)    \n",
    "    sra_table_additional = pandas.concat([sra_table_additional, sra_table_tmp])\n",
    "\n",
    "biosamples = [\"SAMN03299101\",\"SAMN03299100\",\"SAMN03299102\",\"SAMN03105187\",\"SAMN03105186\",\"SAMN04486032\",\"SAMN02739515\",\"SAMN02739515\",\n",
    "              \"SAMN02739513\",\"SAMN02739297\",]\n",
    "sp = \"additional_biosamples\"\n",
    "search_term = get_search_term(species_name=sp, bioprojects=[], biosamples=biosamples)\n",
    "root = fetch_sra_xml(species_name=sp, search_term=search_term, save_xml=True)\n",
    "sra_table_biosamples = sra_xml2table(species_name=sp, xml_root=root)    \n",
    "sra_table_additional = pandas.concat([sra_table_additional, sra_table_biosamples])\n",
    "sra_table_additional.to_csv(\"sra_table_additional_\"+d+\".tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate and save SRA table\n",
    "sra_table_multisp = pandas.read_table(\"sra_table_multisp_\"+d+\".tsv\", sep=\"\\t\", dtype=str)\n",
    "sra_table_additional = pandas.read_table(\"sra_table_additional_\"+d+\".tsv\", sep=\"\\t\", dtype=str)\n",
    "sra_table = pandas.concat([sra_table_multisp, sra_table_additional], axis=0, ignore_index=True)\n",
    "if do_gtex:\n",
    "    sra_table_gtex = pandas.read_table(\"sra_table_GTEx_\"+d+\".tsv\", sep=\"\\t\", dtype=str)\n",
    "    sra_table = pandas.concat([sra_table_multisp, sra_table_gtex, sra_table_additional], axis=0, ignore_index=True)\n",
    "sra_table = sra_table.drop_duplicates(subset=[\"bioproject\", \"biosample\", \"experiment\", \"run\", \"sra_sample\", \"sra_study\"], keep = \"first\")\n",
    "sra_table = sra_table.sort_values(by=[\"scientific_name\", \"bioproject\", \"tissue\", \"treatment\", \"genotype\", \"age\", \"sex\"])\n",
    "sra_table.to_csv(\"sra_table_\"+d+\".tsv\", sep=\"\\t\", index=False)\n",
    "sra_table = pandas.read_table(\"sra_table_\"+d+\".tsv\", sep=\"\\t\", dtype=str)\n",
    "sra_table = sra_table.fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: standardize_orthographical_variants\n",
      "Running: exclude_by_id\n",
      "Running: column_aggregation\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kf/anaconda/lib/python3.5/site-packages/ipykernel_launcher.py:563: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike\n",
      "/Users/kf/anaconda/lib/python3.5/site-packages/pandas/core/indexing.py:1367: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: exclude_entry_with_keyword\n",
      "Running: manual_attribute_filling\n",
      "Running: standardize_scientific_name\n",
      "Running: standardize_tissue\n",
      "Running: standardize_sex\n",
      "Running: exclude_treatment_terms\n",
      "Running: nspot_cutoff\n",
      "Running: remove_redundant_biosample\n",
      "Running: label_sampled_data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kf/anaconda/lib/python3.5/site-packages/ipykernel_launcher.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/kf/anaconda/lib/python3.5/site-packages/ipykernel_launcher.py:495: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/kf/anaconda/lib/python3.5/site-packages/ipykernel_launcher.py:496: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/kf/anaconda/lib/python3.5/site-packages/pandas/core/indexing.py:537: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "409.4720039367676 [sec]\n"
     ]
    }
   ],
   "source": [
    "# Curation\n",
    "start = time.time()\n",
    "sra_table = pandas.read_csv('sra_table_'+d+'.tsv', sep='\\t', header=0, low_memory=False)\n",
    "sra_table_curated = standardize_orthographical_variants(sra_table)\n",
    "sra_table_curated = exclude_by_id(sra_table_curated, bioprojects=[], biosamples=[])\n",
    "sra_table_curated = column_aggregation(sra_table_curated, column_aggregation_list)\n",
    "sra_table_curated = exclude_entry_with_keyword(sra_table_curated)\n",
    "sra_table_curated = manual_attribute_filling(sra_table_curated)\n",
    "sra_table_curated = standardize_scientific_name(sra_table_curated)\n",
    "sra_table_curated = standardize_tissue(sra_table_curated, tissue_aggregation_list)\n",
    "sra_table_curated = standardize_sex(sra_table_curated)\n",
    "sra_table_curated.to_csv('sra_table_column_aggregated_'+d+'.tsv', sep='\\t', index=False)\n",
    "sra_table_curated = exclude_treatment_terms(sra_table_curated, control_term_list)\n",
    "sra_table_curated = nspot_cutoff(sra_table_curated, min_nspots=5000000)\n",
    "sra_table_curated = remove_redundant_biosample(sra_table_curated)\n",
    "sra_table_curated = label_sampled_data(sra_table_curated, suppress_to=20, tissue_selection=tissue_selection)\n",
    "sra_table_curated = reorder_columns(sra_table_curated, omit_misc=True)\n",
    "sra_table_curated.to_csv('sra_table_curated_'+d+'.tsv', sep='\\t', index=False)\n",
    "print(time.time() - start, '[sec]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: make_cross_tables\n",
      "Running: make_cross_tables\n"
     ]
    }
   ],
   "source": [
    "# Cross table\n",
    "sra_qualified_pivot = make_cross_tables(sra_table_curated, n_sp_cutoff=5, suppress_tissue=False, sample_upper=False)\n",
    "sra_qualified_pivot.to_csv('sra_table_pivot_qualified_'+d+'.tsv', sep='\\t')\n",
    "sra_selected_pivot = make_cross_tables(sra_table_curated, n_sp_cutoff=5, suppress_tissue=False, sample_upper=True)\n",
    "sra_selected_pivot.to_csv('sra_table_pivot_selected_'+d+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reduced table\n",
    "sra_table_qualified = sra_table_curated.loc[(sra_table_curated['is_qualified']=='Yes') & (sra_table_curated['is_selected_tissue']=='Yes'),:]\n",
    "sra_table_qualified = sra_table_qualified.loc[(sra_table_qualified['scientific_name']!='Drosophila melanogaster'),:]\n",
    "sra_table_qualified.to_csv('sra_table_qualified_'+d+'.tsv', sep='\\t', index=False)\n",
    "sra_table_reduced = sra_table_curated.loc[(sra_table_curated['is_sampled']=='Yes') & (sra_table_curated['is_selected_tissue']=='Yes'),:]\n",
    "sra_table_reduced.to_csv('sra_table_reduced_'+d+'.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dif\n",
    "\n",
    "sra_dirs = os.listdir(wd_base)\n",
    "#previous_dir = sra_dirs[len(sra_dirs)-2]\n",
    "previous_dir = '2018_5_1'\n",
    "previous_path = wd_base+previous_dir+'/'+'sra_table_pivot_qualified_'+previous_dir+'.tsv'\n",
    "previous_qualified_pivot = pandas.read_csv(previous_path, header=0, index_col=0, sep='\\t')\n",
    "dif_pivot = sra_qualified_pivot.subtract(previous_qualified_pivot, fill_value=0)\n",
    "dif_pivot = dif_pivot.loc[sra_qualified_pivot.index,sra_qualified_pivot.columns]\n",
    "dif_pivot.to_csv('sra_table_pivot_qualified_dif_'+d+'_'+previous_dir+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
